### PR TITLE
ENG-88747 Add converters support in AI instructions

### PR DESCRIPTION
**Pull request by d-krestov**
### `PageSchemaConvertersGuidanceResource`
MCP resource with canonical converter guidance for AI agents working with Freedom UI pages. Covers:
- when to use a converter vs a handler
- `SCHEMA_CONVERTERS` structure (JS object, `usr.*` keys only)
- **async converters** — the runtime automatically handles `instanceof Promise`; `SysSettingsService` is safe (two-layer cache pre-loaded at app startup)
- SDK service access via SCHEMA_DEPS/SCHEMA_ARGS closure
- full `usr.FormatPhoneNumber` template with async/await

### `PageBodyMerger`
Utility for incremental merging of page schema body fragments. Allows AI to send only a delta instead of the full body (re-sending the full body caused "Object vs Array" backend errors). Merge strategies per section:
- `SCHEMA_VIEW_CONFIG_DIFF` — merge by `name`
- `SCHEMA_VIEW_MODEL_CONFIG_DIFF`, `SCHEMA_MODEL_CONFIG_DIFF` — append
- `SCHEMA_HANDLERS` — deduplicate by `request`
- `SCHEMA_CONVERTERS` — deduplicate by key; supports both single- and double-quoted keys

## Renames

| Before | After |
|---|---|
| `PageSchemaSdkCommonGuidanceResource` | `PageSchemaCreatioDevkitCommonGuidanceResource` |
| MCP key `page-schema-sdk-common` | `page-schema-creatio-devkit-common` |

Name aligned with the `@creatio-devkit/common` npm package.

## Updated files

**`GuidanceCatalog`** — added `page-schema-converters` entry; renamed `sdk-common → creatio-devkit-common`.

**Tools** (`GuidanceGetTool`, `PageGetTool`, `PageSyncTool`, `PageUpdateTool`, `ToolContractGetTool`) and **Resources** (`PageSchemaHandlersGuidanceResource`, `PageSchemaValidatorsGuidanceResource`, `PageModificationGuidanceResource`, `PagePrompt`) — cross-references updated to new names.

## Tests

| File | Added |
|---|---|
| `PageToolsTests` | 6 `PageBodyMerger` tests: async converter, nested braces, strings containing structural characters, empty sections, single-quoted keys (merge + overwrite), cross-quote deduplication |
| `SchemaValidationServiceTests` | 2 tests: async arrow function converter, converter with nested curly braces |
| `McpGuidanceResourceTests` | 10 new assertions on `PageSchemaConvertersGuidanceResource` content |
| `GuidanceGetToolTests` | Key names updated |
| E2E | `McpGuidanceResourceE2ETests`, `GuidanceGetToolE2ETests` — keys updated |

### CODE REVIEW

File: clio/Command/McpServer/Resources/GuidanceCatalog.cs [hunk 1]
```diff
@@ -27,14 +27,18 @@ internal static class GuidanceCatalog {
...
+				"Canonical MCP guidance for creating and editing Freedom UI page converters inside raw page schema bodies.",
+				PageSchemaConvertersGuidanceResource.Guide),
 			["page-schema-handlers"] = Create(
 				"page-schema-handlers",
 				"Canonical MCP guidance for creating and editing Freedom UI page handlers inside raw page schema bodies.",
 				PageSchemaHandlersGuidanceResource.Guide),
-			["page-schema-sdk-common"] = Create(
-				"page-schema-sdk-common",
+			["page-schema-creatio-devkit-common"] = Create(
+				"page-schema-creatio-devkit-common",
```
[the remark this answers is not in the document]
Comment (d-krestov) [reply clio/Command/McpServer/Resources/GuidanceCatalog.cs:39]: It's not real case

File: clio/Command/PageBodyMerger.cs [hunk 2]
```diff
@@ -43,15 +51,142 @@ public static string Merge(string currentBody, string incomingBody) {
...
+			// Skip inter-entry whitespace and commas.
+			while (i < inner.Length && (char.IsWhiteSpace(inner[i]) || inner[i] == ',')) {
+				i++;
+			}
+			if (i >= inner.Length) {
+				break;
+			}
+			// Every entry must start with a quoted key.
+			if (inner[i] != '"') {
+				i++;
```
[the remark this answers is not in the document]
Comment (d-krestov) [reply clio/Command/PageBodyMerger.cs]: I fixed it

File: clio.mcp.e2e/McpGuidanceResourceE2ETests.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public async Task McpServer_Should_Advertise_Guidance_Resources() {
 				DataBindingsUri,
 				ExistingAppMaintenanceUri,
 				PageSchemaHandlersUri,
-				PageSchemaSdkCommonUri,
+				PageSchemaCreatioDevkitCommonUri,
```
Comment (Alexandr-Kravchuk) [inline clio.mcp.e2e/McpGuidanceResourceE2ETests.cs:47]: Missing `page-schema-converters` from the expected advertised resources list. AGENTS.md mandates E2E coverage for every new MCP resource: _"Always add or update MCP end-to-end coverage in clio.mcp.e2e for every new or changed MCP tool. This is mandatory."\_ Every other guidance resource (handlers, validators, sdk-common/creatio-devkit-common) is in this list. Without it, a future registration bug silently passes CI.

Suggested fix: add a static field and include it in the assertion:
```csharp
private static readonly string PageSchemaConvertersUri = BuildGuideUri("page-schema-converters");
// in the assertion:
PageSchemaConvertersUri,
```

File: clio.tests/Command/McpServer/GuidanceGetToolTests.cs [hunk 4]
```diff
@@ -197,7 +197,7 @@ public void GuidanceGet_Should_Return_Known_Guide_Names_For_Unknown_Request() {
 				"data-bindings",
 				"existing-app-maintenance",
 				"page-schema-handlers",
-				"page-schema-sdk-common",
+				"page-schema-creatio-devkit-common",
```
Comment (Alexandr-Kravchuk) [inline clio.tests/Command/McpServer/GuidanceGetToolTests.cs:202]: `page-schema-converters` is missing from the `availableGuides` list. When a caller sends an unknown guidance name, this list is the AI's recovery mechanism — omitting the new key means the recovery hint is incomplete. Also, no `GuidanceGet_Should_Return_Page_Schema_Converters_Article` unit test was added; every other registered guidance name has one in this file.

File: clio/Command/PageBodyMerger.cs [hunk 5]
```diff
@@ -43,15 +51,131 @@ public static string Merge(string currentBody, string incomingBody) {
...
+	private static int ScanValueEnd(string s, int i) {
+		int depth = 0;
+		bool inStr = false;
+		char strChar = '"';
+		while (i < s.Length) {
+			char ch = s[i];
+			if (inStr) { i = AdvanceInString(i, ch, strChar, ref inStr); continue; }
+			if (IsStringDelimiter(ch)) { inStr = true; strChar = ch; i++; continue; }
+			if (IsOpenBracket(ch)) { depth++; i++; continue; }
+			if (IsCloseBracket(ch)) { depth--; i++; continue; }
```
Comment (Alexandr-Kravchuk) [inline clio/Command/PageBodyMerger.cs]: **Major: depth can go negative, causing subsequent entries to be silently absorbed**

`IsCloseBracket` fires even when `depth == 0`, driving `depth` negative. Once negative, the guard `depth == 0 && ch == ','` can never fire, so the parser continues past the top-level comma and merges all remaining entries into the current one — meaning every entry after the first nested-bracket value is lost.

Fix: add a depth-floor guard before decrementing:
```csharp
if (IsCloseBracket(ch)) {
    if (depth <= 0) break;
    depth--;
    i++;
    continue;
}
```

File: clio/Command/PageBodyMerger.cs [hunk 6]
```diff
@@ -43,15 +51,131 @@ public static string Merge(string currentBody, string incomingBody) {
...
+	/// at depth 0 could cause a premature entry split. In practice, converter functions are simple
+	/// formatters that do not use regex literals, so this edge case is not expected to occur.
+	/// </remarks>
+	private static List<(string Key, string Entry)> ParseConverterEntries(string inner) {
+		var entries = new List<(string Key, string Entry)>();
+		int i = 0;
+		while (i < inner.Length) {
+			i = SkipSeparators(inner, i);
+			if (i >= inner.Length) break;
+			if (inner[i] != '"' && inner[i] != '\'') { i++; continue; }
```
Comment (Alexandr-Kravchuk) [inline clio/Command/PageBodyMerger.cs]: **Major: unquoted ES6 method-shorthand keys silently erase all converters**

The guard `inner[i] != '"' && inner[i] != '\''` skips every non-quoted character one at a time. When the converter object contains an ES6 method shorthand key — e.g. `toDisplayValue(v) {}` — this branch fires for every character of the key name, then for `(`, `v`, `)`, `{`, ... until the outer loop exhausts the string. The returned `entries` list is empty, so `MergeConvertersRaw` produces `"{}"` and silently discards all converters from both sides.

Creatio converter objects documented in `PageSchemaConvertersGuidanceResource` always use quoted keys (`"crt.ToBoolean"`, `"usr.myConverter"`), so this scenario cannot occur with AI-generated pages. However, any hand-edited page schema that uses method shorthand syntax will lose all converters on the next merge.

Fix: instead of skipping one character, advance past the unrecognised token or return an error:
```csharp
// skip unquoted identifier / keyword
if (char.IsLetterOrDigit(inner[i]) || inner[i] == '_') {
    while (i < inner.Length && inner[i] != ':' && inner[i] != ',') i++;
    continue;
}
i++;
```
Or — if unquoted keys are explicitly out of scope — throw instead of silently skipping so the caller can surface the problem.

**Review by Alexandr-Kravchuk — CHANGES_REQUESTED** [review]
## Review Summary

**Decision: request changes**

The PR adds a well-structured `SCHEMA_CONVERTERS` merge pipeline and a comprehensive MCP guidance resource for Freedom UI converter authoring. The approach is sound and the unit test coverage for the happy path is solid. However, two parser bugs in the new `MergeConvertersRaw` code path can silently corrupt or erase converter data, and the new MCP resource is missing the E2E coverage that AGENTS.md mandates as a blocking requirement.

---

## Findings

### Blockers

**1. AGENTS.md mandatory E2E coverage missing for `page-schema-converters` resource**
File: `clio.mcp.e2e/McpGuidanceResourceE2ETests.cs`, line 43

`McpServer_Should_Advertise_Guidance_Resources` does not include `PageSchemaConvertersUri` in its expected-resources list. The test will pass regardless of whether the new resource is actually advertised by the MCP server. AGENTS.md states: _"Always add or update MCP end-to-end coverage in `clio.mcp.e2e` for every new or changed MCP tool. This is mandatory even when the user does not mention E2E coverage explicitly."_

Fix: add `PageSchemaConvertersUri` to the expected-resources list. Add a new test method that calls the resource URI and asserts the response contains the canonical content (converter table, NON-NEGOTIABLES section, and `SCHEMA_CONVERTERS` structure).

**2. No E2E coverage for `get-guidance name=page-schema-converters`**
File: `clio.mcp.e2e/GuidanceGetToolE2ETests.cs`

The SDK-common key rename is reflected but no new test calls `get-guidance name=page-schema-converters` against a live environment. The mandatory AGENTS.md E2E requirement is unmet.

Fix: add an E2E test that invokes the tool with `name=page-schema-converters` and asserts the response is non-empty and contains expected content.

---

### Majors

**3. `ScanValueEnd` — depth goes negative, subsequent entries silently absorbed**
File: `clio/Command/PageBodyMerger.cs`, line 162 (inline comment posted)

`IsCloseBracket` fires even when `depth == 0`, driving `depth` negative. Once negative, the guard `depth == 0 && ch == ','` can never fire — the parser continues past the top-level comma and absorbs all remaining entries into the current one. Triggered by any converter value whose trailing `}` is the last bracket at depth 0.

Fix: add `if (depth <= 0) break;` guard before decrementing depth.

**4. `ParseConverterEntries` — unquoted ES6 method-shorthand keys silently erase all converters**
File: `clio/Command/PageBodyMerger.cs`, line 111 (inline comment posted)

The guard `inner[i] != '"' && inner[i] != '\''` advances one character at a time for every non-quoted byte. A method shorthand key (`toDisplayValue(v) {}`) exhausts the entire string via single-character skips, returning an empty entries list. `MergeConvertersRaw` then returns `"{}"`, erasing all converters from both sides. Safe for AI-generated pages (all keys are quoted) but destructive for hand-edited schemas.

Fix: advance past the unrecognised token in one pass, or throw with a descriptive message.

**5. Unit test gap — `GuidanceGet_Should_Return_Known_Guide_Names_For_Unknown_Request` missing `page-schema-converters`**
File: `clio.tests/Command/McpServer/GuidanceGetToolTests.cs`, line 200 (inline comment posted)

The `availableGuides` list does not include `"page-schema-converters"`. When an unknown guide name is requested, the error response lists available guides — but `page-schema-converters` will be absent. No dispatch unit test verifies that `get-guidance name=page-schema-converters` resolves to `PageSchemaConvertersGuidanceResource`.

Fix: add `"page-schema-converters"` to `availableGuides`. Add a test asserting the tool dispatches correctly to the new resource.

---

### Minors

**6. Undocumented limitation — nested template literals in `ScanValueEnd`**
File: `clio/Command/PageBodyMerger.cs`

The existing XML doc comment documents the regex-literal limitation. A second undocumented limitation exists: `AdvanceInString` tracks a single `strChar` so nested template literals (`` `outer ${inner}` ``) will drop out of string-tracking mode when the inner `}` closes the interpolation, causing false depth changes. Async converters shown in `PageSchemaConvertersGuidanceResource` use template literals — this edge case is reachable.

Suggested fix: add a note to the XML doc comment; or extend `AdvanceInString` to handle `${` interpolation depth.

---

## Requirement Coverage

No Jira issue or explicit acceptance criteria were provided. Requirements are inferred from the PR title, branch name, and code changes.

| Inferred Requirement | Coverage |
|---|---|
| `SCHEMA_CONVERTERS` section merged by key (incoming wins) | covered |
| New MCP guidance resource `page-schema-converters` registered in `GuidanceCatalog` | covered |
| `page-schema-sdk-common` renamed to `page-schema-creatio-devkit-common` | covered |
| `PageModificationGuidanceResource` updated with converter merge rule | covered |
| MCP E2E coverage for new resource (AGENTS.md mandatory) | **missing** |
| Unit test dispatches `get-guidance name=page-schema-converters` correctly | **missing** |

---

## Confidence Limits

- Requirements are inferred only — no Jira issue or explicit acceptance criteria were available.
- The character-level parser logic was reviewed statically; dynamic execution paths were traced but not run.
- Binary and generated files were not reviewed.
- Coverage of the guidance resource content (~210 lines of documentation text) was spot-checked, not exhaustively validated against the Creatio Freedom UI specification.

File: clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+			         | wrap phone number in tel: link | `crt.ToPhoneLink` | none | no |
+			         | read a specific property from an object attribute | `crt.ToObjectProp` | `prop` (required), `defaultValue` (optional) | no |
+			         | requirement is not covered by rows above | custom `usr.*` converter | custom | yes |
+
+			       Converter binding syntax
+			       - Converters are bound directly in `viewConfigDiff` element property values, NOT in `viewModelConfigDiff`.
+			       - Binding format: `"$AttributeName | converterName"`.
+			       - Example (OOTB, no params): `"caption": "$UsrName | crt.ToBoolean"`.
+			       - Example (custom, no params): `"caption": "$UsrName | usr.ToUpperCase"`.
+			       - Example (OOTB with params): `"caption": "$UsrObject | crt.ToObjectProp:propName"` — params are passed as colon-separated values after the converter name.
```
Comment (Alexandr-Kravchuk) [inline clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs]: **Major: unquoted parameter causes runtime exception at line 57 (and duplicated at line 177)**

The runtime parses each colon-separated token via `_evaluateParam` in `ConverterExpressionInterpreter`. Valid token forms are a quoted string (`'value'`), a number, a boolean, or a binding expression (`$Attr`). A bare identifier (`propName`, `displayValue`) matches none of these and throws:

```
Cannot parse parameter expression: propName
```

Every occurrence of `crt.ToObjectProp` in production page schemas uses quoted string parameters:
```
"$Attr | crt.ToObjectProp : 'displayValue'"
"$Attr | crt.ToObjectProp : 'value'"
```

Fix line 57:
```
- `"caption": "$UsrObject | crt.ToObjectProp:propName"`
+ `"caption": "$UsrObject | crt.ToObjectProp:'propName'"`
```

The same fix applies to line 177 (`crt.ToObjectProp:displayValue` → `crt.ToObjectProp:'displayValue'`).

File: clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+			           }/**SCHEMA_CONVERTERS*/
+			       - The function receives `value` (the current attribute value) and returns the transformed display value.
+			       - Custom converter names MUST use the `usr.` prefix followed by PascalCase (for example `usr.ToUpperCase`, `usr.FormatCurrency`).
+			       - Do NOT name custom converters with the `crt.` prefix — that namespace is reserved for Creatio built-ins.
+
+			       Async converters
+			       - Converters MAY be async. The runtime detects a returned Promise via `instanceof Promise` and awaits it automatically.
+			       - Use `async` arrow function or `async function` syntax — both are valid.
+			       - SDK services from `@creatio-devkit/common` are accessible inside a converter via the SCHEMA_ARGS closure (the same `sdk` argument injected by SCHEMA_DEPS).
+			       - `SysSettingsService` is safe to call in a converter: settings are pre-loaded into a two-layer cache (in-memory Map + IndexedDB) at application startup. Repeated `getByCode` calls return from the cache with no HTTP request.
```
Comment (Alexandr-Kravchuk) [inline clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs]: **Minor: SysSettingsService pre-load claim is overstated**

The guidance says "settings are pre-loaded into a two-layer cache at application startup" without qualification. The two-layer cache (in-memory `Map` + IndexedDB via `CacheManager.executeWithBulkCache`) is real and confirmed — but the preload endpoint (`PreloadSysSettings`) only returns settings the server marks as preloadable. Custom `usr.*` settings are NOT in that set unless explicitly configured in the Creatio admin panel.

For the async converter example that uses `UsrEnablePhoneFormatting`, the first `getByCode` call will make an HTTP request if that setting was never fetched in the session. Subsequent calls are served from the Map cache.

Suggested correction:
```
- settings are pre-loaded into a two-layer cache (in-memory Map + IndexedDB) at application startup. Repeated `getByCode` calls return from the cache with no HTTP request.
+ OOTB Creatio settings are pre-loaded at startup into a two-layer cache (in-memory Map + IndexedDB). Custom `usr.*` settings are NOT pre-loaded — the first `getByCode` call makes one HTTP request; subsequent calls are served from the in-memory cache with no further requests.
```

File: clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			       - OOTB converters (`crt.*`) are built into Creatio — do NOT declare them in `SCHEMA_CONVERTERS`. Only reference them in `viewConfigDiff` bindings.
+			       - Custom converters (`usr.*`) MUST be declared in `SCHEMA_CONVERTERS`.
+			       - Each custom converter entry is a function keyed by its full type name:
+			           converters: /**SCHEMA_CONVERTERS*/{
+			             "usr.ToUpperCase": function(value) {
+			               return value?.toUpperCase() ?? '';
+			             }
+			           }/**SCHEMA_CONVERTERS*/
+			       - The function receives `value` (the current attribute value) and returns the transformed display value.
+			       - Custom converter names MUST use the `usr.` prefix followed by PascalCase (for example `usr.ToUpperCase`, `usr.FormatCurrency`).
```
Comment (Alexandr-Kravchuk) [inline clio/Command/McpServer/Resources/PageSchemaConvertersGuidanceResource.cs:71]: **Minor: PascalCase for `usr.*` converters is a convention, not enforced by runtime**

The guidance states "MUST use the `usr.` prefix followed by PascalCase". In practice, `SchemaConvertersRegistryService.register` calls:
- `checkTypeNameFormat` — requires `Vendor.Name` format (dot separator)
- `checkCrtTypeNamePascalCase` — enforces PascalCase **only for `crt.*`** converters

`usr.myConverter` (camelCase) registers successfully without error. The PascalCase requirement is a team convention, not a runtime constraint. Consider softening to "should use PascalCase" or adding a comment that the platform does not enforce it — so AI consumers understand the rule is stylistic rather than structural.

**Review by Alexandr-Kravchuk — CHANGES_REQUESTED** [review]

**Review by Alexandr-Kravchuk — COMMENTED** [review]
## Follow-up: all findings addressed

All issues raised in the previous review are resolved in commits `136715be` and `bd0245f5`.

### Verified fixes

**Blockers (both resolved)**
- `McpGuidanceResourceE2ETests.cs` — `PageSchemaConvertersUri` added to the advertised-resources list; `McpServer_Should_Return_Page_Schema_Converters_Guidance` test added.
- `GuidanceGetToolE2ETests.cs` — `GuidanceGet_Should_Return_Page_Schema_Converters_Guide` E2E test added.

**Majors (all resolved)**
- `ScanValueEnd` depth guard — `if (depth <= 0) break;` added before decrementing; regression test `PageBodyMerger_Should_Not_Absorb_Next_Entry_After_Nested_Brackets` added.
- `ParseConverterEntries` unquoted key — refactored into `ReadQuotedEntry` / `SkipUnquotedEntry` / `SkipUnquotedKey`; regression test `PageBodyMerger_Should_Not_Corrupt_Quoted_Entry_After_Unquoted_Key` added.
- `GuidanceGetToolTests.cs` — `"page-schema-converters"` added to `availableGuides`; `GuidanceGet_Should_Return_Page_Schema_Converters_Article` dispatch test added.

**Minors (all resolved)**
- `crt.ToObjectProp` parameter syntax — fixed at L57 and L177: `:'propName'` / `:'displayValue'` (single-quoted).
- `SysSettingsService` cache claim — corrected to distinguish OOTB pre-loaded settings from custom `usr.*` settings.
- `PascalCase` wording — softened to a team convention note; `SkipUnquotedKey` `remarks` documented.

The `remarks` block in `ScanValueEnd` documenting the template-literal limitation was also added as requested.

**Review by Alexandr-Kravchuk — APPROVED** [review]

**Review by a-trofimchuk — APPROVED** [review]
